rfcs/037-api-faceting-principles/README.md

```diff
@@ -0,0 +1,146 @@
+
+## Principles
+
+**1. Filters are named by the JSON paths of the attribute they use**
+
+For example, given a display document that looks like:
+
+```json
+{
+  "a": {
+    "b": [
+      {
+        "c": "thing1"
+      },
+      {
+        "c": "thing2"
+      }
+    ]
+  }
+}
+```
+
+Then a document filter that filtered by the values of the `c` property would use a query string like `a.b.c`. In this example, filtering for `c` equal to `thing2` might look like:
+
+```
+http://host.name/path/docs?a.b.c=thing2
+```
+
+**2. Aggregations are always paired with identically named filters**
+
+It is not strictly necessary that all filters have aggregations, but all aggregations must be present alongside an identically named filter for the property that is being aggregated upon. This document will refer to these as "paired" filters and aggregations. For the above example, an aggregation on the values of the  `c` property would be used like this:
+
+```
+http://host.name/path/docs?aggregations=a.b.c
+```
+
+**3. Aggregations are returned in an `aggregations` field, with the same name by which they were requested** 
+
+This means JSON paths are still represented as strings, rather than being expanded. For example, the response to the previous example would include at the top level
+
+```
+{
+  ...,
+  "aggregations": {
+    "a.b.c": {
+       "buckets": [
+         ...
+       ]
+    }
+  }
+}
+```
+
+
+
+**4. Aggregation buckets contain a `data` field of the same type as the aggregated property's parent entity**
+
+That is to say, while we aggregate on a specific string field (for example, an ID or a label), we want to return the full entity that contains the field. This is easiest to explain by example. Given the following display documents:
+
+```json
+{
+  "a": {
+    "b": [
+      {
+        "id": "id1",
+        "label": "Thing 1"
+      },
+      {
+        "id": "id2",
+        "label": "The 2nd thing"
+      }
+    ]
+  }
+}
+```
+
+Then if we aggregate on the labels like this:
+
+```
+http://host.name/path/docs?aggregations=a.b.label
+```
+
+Then our response buckets will look something like this:
+
+```
+{
+  "data": {
+    "id": "id1",
+    "label": "Thing 1"
+  },
+  "count": 1234
+}
+```
+
+**5. When a filter and its paired aggregation are both applied, that aggregation's buckets are not filtered**
+
+Conversely, filters do apply to the buckets of all aggregations other than the paired aggregation. This initially confusing requirement is necessary because - for mutually exclusive values - application of the filter to the aggregation buckets will remove all but the selected bucket, thus removing the ability of the interface to show other options for the given filter. This is explained further in the [Elasticsearch documentation for `post_filter`](https://www.elastic.co/guide/en/elasticsearch/reference/current/filter-search-results.html#post-filter).
```

Comment:
I agree with the conclusion, but I think this would be better explained with an example. You also need to cover the case where values aren't mutually exclusive.

AIUI, our filters use OR rather than AND – e.g. if you apply the filter `?language=english,french,german`, it will find

```
(documents with "english") ∩ (documents with "french") ∩ (documents with "german")
```

When we present a UI for this, we want to be able to show all three options. If I select one, the other two should be available, so I can add that to expand my search.

Reply:
Hmmm, I will explain this better and add an example. The case where the values aren't mutually exclusive is not relevant here - the problem only exists for when the values are mutually exclusive. I will clarify this though.